docs/antrea-network-policy.md

```diff
@@ -1221,16 +1300,33 @@ share the `view` ClusterRole to a wider range of subjects to allow them to read
 the policies that may affect their workloads.
 Similar RBAC is applied to the ClusterGroup resource.
 
-## Notes
+## Notes and constraints
 
 - There is a soft limit of 20 on the maximum number of Tier resources that are
   supported. But for optimal performance, it is recommended that the number of
   Tiers in a cluster be less than or equal to 10.
 - In order to reduce the churn in the agent, it is recommended to set the policy
-  priority within the range 1.0 to 100.0.
+  priority (acnp/anp.spec.priority) within the range 1.0 to 100.0.
 - The v1alpha1 policy CRDs support up to 10,000 unique priorities at policy level,
-  and up to 50,000 unique priorities at rule level, across all tiers except for
-  the "baseline" tier. For any two policy rules, their rule level priorities are only
-  considered equal if they share the same tier, and have the same policy priority
-  as well as rule priority.
-- For the "baseline" tier, the max supported unique priorities (at rule level）is 150.
+  and up to 50,000 unique priorities at rule level, across all Tiers except for
+  the "baseline" Tier. For any two Antrea-native policy rules, their rule level
+  priorities are only considered equal if their policy objects share the same Tier
+  and have the same policy priority, plus the rules themselves are of the same rule
+  priority (rule priority is the sequence number of the rule within the policy's
+  ingress or egress section).
+- For the "baseline" Tier, the max supported unique priorities (at rule level) is 150.
+- If there are multiple Antrea-native policy rules created at the same rule-level
+  priority (same policy Tier, policy priority and rule priority), who happen to select
+  overlapping traffic patterns but have conflicting rule actions (e.g.`Allow`v.s.`Deny`),
+  the behavior of such traffic will be undeterministic. For example, consider two
+  AntreaNetworkPolicies created in the same Namespace and Tier with the same policy
+  priority. The first policy applies to all `app=web` Pods in the Namespace and has only
+  one ingress rule to `Deny` all traffic from `role=dev` Pods. The other policy also
+  applies to all `app=web` Pods in the Namespace and has only one ingress rule, which is
+  to `Allow` all traffic from `app=client` Pods. Those two ingress rules might not always
+  conflict, but in case a Pod with both `app=client` and `role=dev` labels initiate
+  traffic towards the web Pods in the Namespace, both rules will be matched at the same
+  priority with conflicting actions. It will be the policy writer's responsibility to
+  identify such ambiguities in rule definitions and avoid potential undeterministic
+  rule enforcement results. In general, it is recommended that rules of conflicting
+  actions are always created at different rule-level priorities.
```

Comment:
I think this is not strong enough. IMO, it should be:

> In general, it is recommended not to create rules with conflicting actions in policy resources with the same priority, if these rules may apply to the same traffic, as the behavior could be undeterministic.

And actually I think this should be the first sentence of that bullet list item (and in bold), and you can get into the details after if needed, as Jianjun suggested. The whole concept of rule level priority for rules which are defined across policies with the same priority, is an implementation detail as far as I am concerned. I don't think it is really needs to be a part of the API contract. Inserting a new rule in one of the policies could change rule enforcement order, in ways which are surprising to users.

Reply:
Done